File: src/deploy/yolov5/cpp/yolov5_detector.cpp
/*
 * Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

#include "yolov5_detector.h"

#include <algorithm>
#include <chrono>
#include <cmath>
#include <memory>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

#include "common.h"
#include "vision_model_config.h"
#include "vision_model_factory.h"

namespace vision_deploy {

namespace {
inline float sigmoid(float x) {
    return 1.0f / (1.0f + std::exp(-x));
}
}  // namespace

std::unique_ptr<vision_core::BaseModel> YOLOv5Detector::create(const YAML::Node& config, bool lazy_load) {
    std::string model_path = vision_core::yaml_utils::getString(config, "model_path");
    if (model_path.empty()) {
        throw std::runtime_error("model_path not found in config for YOLOv5Detector");
    }

    YAML::Node default_params = config["default_params"];
    if (!default_params) {
        throw std::runtime_error("default_params not found in config for YOLOv5Detector");
    }

    float conf_threshold = vision_core::yaml_utils::getFloat(default_params, "conf_threshold", 0.25f);
    float iou_threshold = vision_core::yaml_utils::getFloat(default_params, "iou_threshold", 0.45f);
    int num_threads = vision_core::yaml_utils::getInt(default_params, "num_threads", 4);
    std::string provider = vision_core::yaml_utils::getProvider(config);

    return std::make_unique<YOLOv5Detector>(
        model_path, conf_threshold, iou_threshold, num_threads, lazy_load, provider);
}

YOLOv5Detector::YOLOv5Detector(const std::string& model_path,
                                float conf_threshold,
                                float iou_threshold,
                                int num_threads,
                                bool lazy_load,
                                const std::string& provider)
    : BaseModel(model_path, lazy_load),
        conf_threshold_(conf_threshold),
        iou_threshold_(iou_threshold),
        num_threads_(num_threads),
        provider_(provider) {
    if (!lazy_load) {
        load_model();
    }
}

void YOLOv5Detector::load_model() {
    init_session(num_threads_, provider_);
    model_loaded_ = true;
}

cv::Mat YOLOv5Detector::preprocess(const cv::Mat& image) {
    if (image.empty()) {
        throw std::runtime_error("Input image is empty");
    }

    ensure_model_loaded();

    int input_width = static_cast<int>(input_shape_[3]);
    int input_height = static_cast<int>(input_shape_[2]);

    cv::Mat padded = vision_common::letterbox(image, std::make_pair(input_height, input_width));
    return cv::dnn::blobFromImage(
        padded, 1.0f / 255.0f, cv::Size(input_width, input_height), cv::Scalar(0, 0, 0), true, false, CV_32F);
}

std::vector<vision_common::Result> YOLOv5Detector::detect(const cv::Mat& image) {
    ensure_model_loaded();
    reset_runtime_profile();
    const auto t0 = std::chrono::steady_clock::now();

    const cv::Size orig_size = image.size();

    const auto t_pre0 = std::chrono::steady_clock::now();
    cv::Mat input_tensor = preprocess(image);
    const auto t_pre1 = std::chrono::steady_clock::now();
    set_runtime_preprocess_ms(std::chrono::duration<double, std::milli>(t_pre1 - t_pre0).count());

    const auto t_infer0 = std::chrono::steady_clock::now();
    std::vector<Ort::Value> outputs = run_session(input_tensor);
    const auto t_infer1 = std::chrono::steady_clock::now();
    set_runtime_model_infer_ms(std::chrono::duration<double, std::milli>(t_infer1 - t_infer0).count());

    const auto t_post0 = std::chrono::steady_clock::now();
    std::vector<vision_common::Result> results = postprocess(outputs, orig_size);
    const auto t_post1 = std::chrono::steady_clock::now();
    set_runtime_postprocess_ms(std::chrono::duration<double, std::milli>(t_post1 - t_post0).count());

    const auto t1 = std::chrono::steady_clock::now();
    set_runtime_total_ms(std::chrono::duration<double, std::milli>(t1 - t0).count());

    return results;
}

std::vector<vision_core::ModelCapability> YOLOv5Detector::get_capabilities() const {
    return {
        vision_core::ModelCapability::kImageInput,
        vision_core::ModelCapability::kDraw};
}

std::vector<vision_common::Result> YOLOv5Detector::postprocess(
    std::vector<Ort::Value>& outputs,
    const cv::Size& orig_size) {
    if (outputs.empty()) {
        return {};
    }

    const int input_height = static_cast<int>(input_shape_[2]);
    const int input_width = static_cast<int>(input_shape_[3]);

    Ort::Value& out0 = outputs[0];
    auto info = out0.GetTensorTypeAndShapeInfo();
    std::vector<int64_t> dims = info.GetShape();
    if (dims.size() < 2) {
        throw std::runtime_error("Unexpected YOLOv5 output shape (dims < 2)");
    }

    int64_t num_boxes = 0;
    int64_t features = 0;
    bool layout_cn = false;
    if (dims.size() == 3) {
        const int64_t dim1 = dims[1];
        const int64_t dim2 = dims[2];
        if ((dim1 <= 256 && dim2 > 256) || dim1 < dim2) {
            features = dim1;
            num_boxes = dim2;
            layout_cn = true;
        } else {
            num_boxes = dim1;
            features = dim2;
            layout_cn = false;
        }
    } else if (dims.size() == 2) {
        num_boxes = dims[0];
        features = dims[1];
    } else {
        features = dims.back();
        num_boxes = 1;
        for (size_t i = 0; i + 1 < dims.size(); ++i) {
            if (i == 0 && dims[0] == 1) {
                continue;
            }
            num_boxes *= dims[i];
        }
    }

    if (features < 5) {
        throw std::runtime_error("Unexpected YOLOv5 output features (< 5)");
    }

    const bool has_objectness = (features >= 85);
    const int64_t cls_start = has_objectness ? 5 : 4;
    const int64_t num_classes = features - cls_start;

    const float* data = out0.GetTensorData<float>();

    auto get_val = [&](int64_t box_idx, int64_t ch_idx) -> float {
        if (layout_cn) {
            return data[ch_idx * num_boxes + box_idx];
        }
        return data[box_idx * features + ch_idx];
    };

    bool need_sigmoid = false;
    bool need_scale_xywh = false;
    const int64_t probe_n = std::min<int64_t>(num_boxes, 200);
    float min_conf = 1e9f;
    float max_conf = -1e9f;
    float max_xywh = 0.0f;
    for (int64_t i = 0; i < probe_n; ++i) {
        for (int k = 0; k < 4; ++k) {
            max_xywh = std::max(max_xywh, std::fabs(get_val(i, k)));
        }
        if (has_objectness) {
            const float obj_raw = get_val(i, 4);
            min_conf = std::min(min_conf, obj_raw);
            max_conf = std::max(max_conf, obj_raw);
        }
        if (num_classes > 0) {
            const float cls_raw = get_val(i, cls_start);
            min_conf = std::min(min_conf, cls_raw);
            max_conf = std::max(max_conf, cls_raw);
        }
    }
    if (max_conf > 1.5f || min_conf < -0.1f) {
        need_sigmoid = true;
    }
    if (max_xywh <= 1.5f) {
        need_scale_xywh = true;
    }

    std::vector<vision_common::Result> candidates;
    candidates.reserve(static_cast<size_t>(num_boxes));

    for (int64_t i = 0; i < num_boxes; ++i) {
        float obj = 1.0f;
        if (has_objectness) {
            obj = get_val(i, 4);
            if (need_sigmoid) {
                obj = sigmoid(obj);
            }
            if (obj <= conf_threshold_) {
                continue;
            }
        }

        float best_conf = 0.0f;
        int best_cls = -1;
        for (int64_t c = 0; c < num_classes; ++c) {
            float cls = get_val(i, cls_start + c);
            if (need_sigmoid) {
                cls = sigmoid(cls);
            }
            const float conf = has_objectness ? (cls * obj) : cls;
            if (conf > best_conf) {
                best_conf = conf;
                best_cls = static_cast<int>(c);
            }
        }

        if (best_cls < 0 || best_conf <= conf_threshold_) {
            continue;
        }

        float xywh[4] = {get_val(i, 0), get_val(i, 1), get_val(i, 2), get_val(i, 3)};
        if (need_scale_xywh) {
            xywh[0] *= static_cast<float>(input_width);
            xywh[1] *= static_cast<float>(input_height);
            xywh[2] *= static_cast<float>(input_width);
            xywh[3] *= static_cast<float>(input_height);
        }

        float xyxy[4];
        vision_common::xywh2xyxy(xywh, xyxy);

        vision_common::Result r;
        r.x1 = xyxy[0];
        r.y1 = xyxy[1];
        r.x2 = xyxy[2];
        r.y2 = xyxy[3];
        r.score = best_conf;
        r.label = best_cls;
        candidates.push_back(r);
    }

    if (candidates.empty()) {
        return {};
    }

    std::vector<cv::Rect2f> boxes;
    std::vector<float> scores;
    boxes.reserve(candidates.size());
    scores.reserve(candidates.size());
    for (const auto& cand : candidates) {
        boxes.emplace_back(cand.x1, cand.y1, cand.x2 - cand.x1, cand.y2 - cand.y1);
        scores.push_back(cand.score);
    }

    std::vector<int> keep_indices = vision_common::nms(boxes, scores, iou_threshold_);
    std::vector<vision_common::Result> results;
    results.reserve(keep_indices.size());
    for (int idx : keep_indices) {
        results.push_back(candidates[static_cast<size_t>(idx)]);
    }

    const cv::Size input_shape(input_width, input_height);
    for (auto& r : results) {
        float coords[4] = {r.x1, r.y1, r.x2, r.y2};
        vision_common::scale_coords(input_shape, coords, orig_size);
        r.x1 = coords[0];
        r.y1 = coords[1];
        r.x2 = coords[2];
        r.y2 = coords[3];
    }

    return results;
}

static vision_core::ModelRegistrar<YOLOv5Detector> registrar("YOLOv5Detector");

}  // namespace vision_deploy

File: src/deploy/yolov5/python/yolov5_detector.py
# Copyright 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
#
# SPDX-License-Identifier: Apache-2.0

"""
YOLOv5 Object Detector Implementation
"""

import cv2
import numpy as np
import onnxruntime as ort
import spacemit_ort  # noqa: F401 - register SpaceMITExecutionProvider
from typing import List, Tuple, Optional, Dict, Any
from pathlib import Path

from core import BaseModel
from common import letterbox, nms, xywh2xyxy, scale_coords
from core.python.vision_model_exceptions import (
    ModelNotFoundError,
    ModelLoadError,
    ModelInferenceError,
)


class YOLOv5Detector(BaseModel):
    """
    YOLOv5 object detection model.

    Supports detection with various input sizes and configurations.
    Handles both sigmoid and non-sigmoid outputs automatically.
    """

    def __init__(self, model_path: str,
                 conf_threshold: float = 0.25,
                 iou_threshold: float = 0.45,
                 num_threads: int = 4,
                 **kwargs):
        """
        Initialize YOLOv5 detector.

        Args:
            model_path: Path to ONNX model file
            conf_threshold: Confidence threshold for filtering detections
            iou_threshold: IoU threshold for NMS
            num_threads: Number of threads for ONNX Runtime
            **kwargs: Additional parameters
        """
        self.conf_threshold = conf_threshold
        self.iou_threshold = iou_threshold
        self.num_threads = num_threads
        super().__init__(model_path, **kwargs)

    def _load_model(self, **kwargs):
        """Load YOLOv5 ONNX model."""
        model_path_obj = Path(self.model_path)
        if not model_path_obj.exists():
            raise ModelNotFoundError(
                model_path=self.model_path,
                message=f"YOLOv5 模型文件不存在: {self.model_path}"
            )

        session_options = ort.SessionOptions()
        session_options.intra_op_num_threads = self.num_threads

        try:
            providers = kwargs.get('providers', ['SpaceMITExecutionProvider'])
            self.session = ort.InferenceSession(
                self.model_path,
                sess_options=session_options,
                providers=providers
            )
        except Exception as e:
            raise ModelLoadError(
                model_path=self.model_path,
                reason=str(e),
                message=f"YOLOv5 模型加载失败: {self.model_path} - {e}"
            )

        try:
            self.input_name = self.session.get_inputs()[0].name
            self.output_names = [output.name for output in self.session.get_outputs()]
            self.input_shape = tuple(self.session.get_inputs()[0].shape[2:4])
        except Exception as e:
            raise ModelLoadError(
                model_path=self.model_path,
                reason=f"无法获取模型输入/输出信息: {e}",
                message=f"YOLOv5 模型初始化失败: {self.model_path} - {e}"
            )

    def preprocess(self, image: np.ndarray) -> Tuple[np.ndarray, float, Tuple[float, float]]:
        """
        Preprocess image for YOLOv5 inference.

        Args:
            image: Input image in BGR format

        Returns:
            Tuple of (preprocessed_tensor, ratio, pad)
        """
        img_resized, r, (dw, dh) = letterbox(image, self.input_shape)

        # BGR to RGB
        img_rgb = cv2.cvtColor(img_resized, cv2.COLOR_BGR2RGB)

        # Normalize to [0, 1]
        img_normalized = img_rgb.astype(np.float32) / 255.0

        # HWC to CHW
        img_chw = np.transpose(img_normalized, (2, 0, 1))

        # Add batch dimension
        img_batch = np.expand_dims(img_chw, axis=0)

        return img_batch, r, (dw, dh)

    @staticmethod
    def _sigmoid(x):
        """Sigmoid activation function."""
        return 1.0 / (1.0 + np.exp(-x))

    def postprocess(
        self,
        outputs: List[np.ndarray],
        orig_shape: Tuple[int, int],
        ratio: float,
        pad: Tuple[float, float],
        **kwargs,
    ) -> List[Dict[str, Any]]:
        """
        Postprocess YOLOv5 outputs.

        Args:
            outputs: Raw model outputs
            orig_shape: Original image shape (H, W)
            ratio: Scale ratio from letterbox
            pad: Padding (dw, dh) from letterbox
            **kwargs: Additional parameters

        Returns:
            List of detection dictionaries
        """
        if not outputs or len(outputs) == 0:
            return []

        output = outputs[0]

        # Handle different output shapes
        if output.ndim == 3:
            # Shape: [1, num_boxes, features] or [1, features, num_boxes]
            if output.shape[1] <= 256 and output.shape[2] > 256:
                # Layout: [1, features, num_boxes] -> transpose
                output = output.transpose(0, 2, 1)
            output = output[0]  # Remove batch dimension
        elif output.ndim == 2:
            # Shape: [num_boxes, features]
            pass
        else:
            raise ValueError(f"Unexpected YOLOv5 output shape: {output.shape}")

        num_boxes, features = output.shape

        if features < 5:
            raise ValueError(f"Unexpected YOLOv5 output features: {features}")

        # Detect if model has objectness score (85+ features for COCO)
        has_objectness = (features >= 85)
        cls_start = 5 if has_objectness else 4
        num_classes = features - cls_start

        # Auto-detect if sigmoid is needed by probing first 200 boxes
        probe_n = min(num_boxes, 200)
        probe_data = output[:probe_n]

        need_sigmoid = False
        need_scale_xywh = False

        # Check confidence range
        if has_objectness:
            obj_vals = probe_data[:, 4]
            if obj_vals.max() > 1.5 or obj_vals.min() < -0.1:
                need_sigmoid = True

        if num_classes > 0:
            cls_vals = probe_data[:, cls_start:]
            if cls_vals.max() > 1.5 or cls_vals.min() < -0.1:
                need_sigmoid = True

        # Check if xywh needs scaling
        xywh_vals = probe_data[:, :4]
        if np.abs(xywh_vals).max() <= 1.5:
            need_scale_xywh = True

        # Extract objectness scores
        if has_objectness:
            obj_scores = output[:, 4]
            if need_sigmoid:
                obj_scores = self._sigmoid(obj_scores)
        else:
            obj_scores = np.ones(num_boxes, dtype=np.float32)

        # Extract class scores
        class_scores = output[:, cls_start:]
        if need_sigmoid:
            class_scores = self._sigmoid(class_scores)

        # Compute final confidence scores
        if has_objectness:
            confidences = class_scores * obj_scores[:, np.newaxis]
        else:
            confidences = class_scores

        # Get best class for each box
        class_ids = np.argmax(confidences, axis=1)
        scores = confidences[np.arange(num_boxes), class_ids]

        # Filter by confidence threshold
        mask = scores > self.conf_threshold
        if not mask.any():
            return []

        scores = scores[mask]
        class_ids = class_ids[mask]
        boxes_xywh = output[mask, :4]

        # Scale xywh if needed
        if need_scale_xywh:
            boxes_xywh[:, 0] *= self.input_shape[1]  # x * width
            boxes_xywh[:, 1] *= self.input_shape[0]  # y * height
            boxes_xywh[:, 2] *= self.input_shape[1]  # w * width
            boxes_xywh[:, 3] *= self.input_shape[0]  # h * height

        # Convert xywh to xyxy
        boxes_xyxy = xywh2xyxy(boxes_xywh)

        # Apply NMS
        keep_indices = nms(boxes_xyxy, scores, self.iou_threshold)

        if len(keep_indices) == 0:
            return []

        boxes_xyxy = boxes_xyxy[keep_indices]
        scores = scores[keep_indices]
        class_ids = class_ids[keep_indices]

        # Scale boxes back to original image size
        input_size = (self.input_shape[1], self.input_shape[0])  # (W, H)
        boxes_xyxy = scale_coords(input_size, boxes_xyxy, orig_shape)

        # Format results
        results = []
        for box, score, class_id in zip(boxes_xyxy, scores, class_ids):
            results.append({
                'bbox': box.tolist(),
                'confidence': float(score),
                'class_id': int(class_id)
            })

        return results

    def infer(self, image: np.ndarray) -> List[Dict[str, Any]]:
        """
        Run YOLOv5 inference on an image.

        Args:
            image: Input image in BGR format

        Returns:
            List of detection dictionaries with keys:
            - bbox: [x1, y1, x2, y2] in original image coordinates
            - confidence: detection confidence score
            - class_id: class index
        """
        self._ensure_model_loaded()

        orig_shape = image.shape[:2]

        try:
            input_tensor, ratio, pad = self.preprocess(image)
        except Exception as e:
            raise ModelInferenceError(
                model_name="YOLOv5",
                reason=f"图像预处理失败: {e}",
                message=f"YOLOv5 预处理错误: {e}"
            )

        try:
            outputs = self.session.run(self.output_names, {self.input_name: input_tensor})
        except Exception as e:
            raise ModelInferenceError(
                model_name="YOLOv5",
                reason=f"模型推理失败: {e}",
                message=f"YOLOv5 推理错误: {e}"
            )

        try:
            detections = self.postprocess(outputs, orig_shape, ratio, pad)
        except Exception as e:
            raise ModelInferenceError(
                model_name="YOLOv5",
                reason=f"结果后处理失败: {e}",
                message=f"YOLOv5 后处理错误: {e}"
            )

        return detections
